Use fourcipp

davidrudlstorfer opened the pull request:
Utilize FourCIPP in MeshPy

davidrudlstorfer commented:
@isteinbrecher this is a heavy WIP as stated above and many tests are still failing.

I think it is a good idea that we review/discuss/talk about this in an early stage so we are on the same track. This would be a first rough proposal on how to use fourcipp in MeshPy.

I know that I need to update a lot in the model importer and mesh to dict comparison.

Let me know if we want to do a short review here or talk via Zoom etc.

Thanks in advance

davidrudlstorfer commented:
One note:

With https://github.com/4C-multiphysics/fourcipp/pull/39 fourcipp is now completely compatible with all of our tested systems, i.e., we can now install everything seamlessly on macOS, Windows and Ubuntu across all of our Python versions. I think that's nice:)

isteinbrecher reviewed (commented):
Thanks a lot for the changes @davidrudlstorfer, and very nice that fourcipp supports all supported versions / OS that we need.

I had a quick look (not at the tests and the workflows) and think this is already a very good first shot, nice to see how much code can be removed!

We can definitely have a talk this week, I am not sure when is good for me, I will let you know on slack.

davidrudlstorfer commented on src/meshpy/four_c/element_beam.py:
```diff
@@ -133,32 +132,28 @@ def __init__(self, *, weak=True, rotvec=True, is_fad=True, **kwargs):
-        )
-
-        return [string_dat]
+        # TODO add FAD key? How is this set in the dict with no value in fourcipp?
```
This now works - I was unsure on how the syntax on the 4C side was and how this was converted into our "FourCIPP dict".

davidrudlstorfer commented:
@isteinbrecher first of all thanks for your first round of review and the helpful commands. This turned out to be more involved than I initially thought but I am pretty happy with the current state.  Nearly all tests are currently working. Two tests are still failing because of an bug I found in fourcipp (should be resolved here https://github.com/4C-multiphysics/fourcipp/pull/42)  The remaining 9 tests are connected to nurbs. I am currently unsure on how to proceed with the Nurbs because I have absolutely no clue and the present input files are not compatible with 4C as far as I can tell because FourCIPP has some problems with it. Maybe you can quickly fix the nurbs stuff - but lets talk about this tomorrow together.  I’ve noted the following overall points which must be completed before merging this PR: 
- nurbs: fix dump_to_list functions to be compatible with fourcipp

- fourcipp: Once the nurbs stuff is fixed we must remove the function is_fourcipp_available

After these things are fixed I would merge this and I noted the following clean-ups after this PR

- can we finally delete the tutorial or should be wait longer? 
- fourcipp: a dumper for numpy types (arrays, floats, ints, booleans) will be added soon. Therefore, I would keep the int/float/list comparison for now in the dump_to_list functions and remove it soon from MeshPy.

- move all dump_to_list logic to the input file (similar to boundary_condition, function, etc…)

- move element_type_to_four_c_string to a more central place, i.e., out of core. This should be easily possible once the dump_to_list functions are in the input file (currently a circular import restricts me from moving this)

- then we can also fix the current hackery in the solidrigidsphere for free (here also a circular import hinders me from working further on this)

- currently the overwrite option is gone, this is not as easily implementable because one does actually not know what should be overwritten? If there is a plausible solution to this we can add it to FourCIPP

- Finally we can improve and simplify (by a lot) the model importer

- And finally, finally we can improve and simplify the mesh to dict conversion (by a lot)

This took me longer than anticipated because I also needed to adjust a lot of reference files because they actually were not compatible with 4C. I think if we fix the nurbs stuff this should be mergable. The remaining points I noted should actually be pretty straight forward. But if I add them to this PR it would explode and would be pretty bad to review.  I am looking forward to our meeting tomorrow:)

davidrudlstorfer commented:
@isteinbrecher this is ready for review :partying_face: 

I've already incorporated all of your raised points above and I've also added comments into the code base where we convert from numpy data types to standard python data types for FourCIPP.

I am looking forward to your review - and hopefully we can merge this soon:)

davidrudlstorfer commented:
I am closing this.

#354 contains the identical code changes + the new type converter from FourCIPP. With this we do no longer need to convert our types in MeshPy.

:rocket:
